Mathlib/Algebra/Category/Ring/Center.lean

```diff
@@ -0,0 +1,93 @@
+/-
+Copyright (c) 2025 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang, Junyan Xu
+-/
+import Mathlib.CategoryTheory.Adjunction.Limits
+import Mathlib.Algebra.Category.ModuleCat.Abelian
+import Mathlib.Algebra.Category.ModuleCat.ChangeOfRings
+import Mathlib.Algebra.Small.Module
+
+/-!
+# A categorical description of the center of a ring
+
+In this file we prove that the center of a ring `R` is isomorphic to `End (𝟭 R-Mod)` the
+endomorphism ring of the identity functor on the category of `R`-modules. Consequently, the ring
+structure of a commutative ring is completely determined by its module category.
+
+## Main results
+
+- `Subring.centerEquivEndIdFunctor`: the center of a ring `R` is isomorphic to `End (𝟭 R-Mod)`.
+- `RingEquiv.ofModuleCatEquiv`: if two commutative rings have equivalent module categories, they are
+  isomorphic as rings.
+
+-/
+
+universe v v'
+
+variable (R : Type*) [Ring R]
+
+open CategoryTheory
+
+noncomputable abbrev test [Small.{v} R] : ModuleCat.{v} R ≌ ModuleCat.{v} (Shrink.{v} R) :=
+  ModuleCat.restrictScalarsEquivalenceOfRingEquiv (Shrink.ringEquiv R)
+
+variable (M : ModuleCat R)
+
+set_option synthInstance.maxHeartbeats 100000 in
+/--
+For any ring `R`, the center of `R` is isomorphic to `End (𝟭 (ModuleCat R))`, the endomorphism ring
+of the identity functor on the category of `R`-modules.
+-/
+@[simps]
+noncomputable def Subring.centerEquivEndIdFunctorAux [Small.{v} R] :
+    center (Shrink.{v} R) ≃+* End (𝟭 (ModuleCat.{v} (Shrink.{v} R))) where
+  toFun r :=
+    { app M := r • 𝟙 M }
+  invFun f := centerToMulOpposite.symm <| centerCongr
+    ((ModuleCat.of _ (Shrink.{v} R)).endRingEquiv.trans
+      ((Module.moduleEndSelf (Shrink.{v} R))).symm)
+    ⟨f.app _, mem_center_iff.mpr fun g ↦ (f.naturality _).symm⟩
+  left_inv r := Subtype.ext <| show r.1 • (1 : Shrink R) = r.1 by simp
+  right_inv f := by
+    apply NatTrans.ext
+    ext M (m : M)
+    simpa using
+      congr($(f.naturality (X := .of _ (Shrink.{v} R)) (Y := .of _ M) <|
+        ModuleCat.ofHom <| LinearMap.toSpanSingleton _ M m).hom (1 : Shrink.{v} R)).symm
+  map_mul' x y := by
+    apply NatTrans.ext
+    ext M (m : M)
+    simpa using mul_smul x.1 y.1 m
+  map_add' x y := by
+    apply NatTrans.ext
+    ext M (m : M)
+    simpa using add_smul x.1 y.1 m
```

Comment:
These proofs are hiding some heavy rfls, as witnessed by the `erw`s below
```suggestion
    simp
    erw [End.mul_def] -- why is this simp lemma not firing?
    simp
    exact mul_smul x y m
  map_add' x y := by
    apply NatTrans.ext
    ext M (m : M)
    simp
    erw [NatTrans.app_add] -- why is this simp lemma not firing?
    simp
    exact add_smul x y m
```

Reply:
```
erw [NatTrans.app_add] -- why is this simp lemma not firing?
```
Strangely, this does not require `erw`, `rw` actually works, but `simp` does not fire nonetheless

Reply:
Probably [this issue](https://leanprover.zulipchat.com/#narrow/channel/270676-lean4/topic/.60simp.60.20is.20sometimes.20less.20powerful.20than.20.60rw.60/near/497521128).

Reply:
changed